src/theory/quantifiers/sygus/example_eval_cache.cpp

```diff
@@ -0,0 +1,123 @@
+/*********************                                                        */
+/*! \file example_eval_cache.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief
+ **
+ **/
+#include "theory/quantifiers/sygus/example_eval_cache.h"
+
+#include "theory/quantifiers/sygus/example_min_eval.h"
+#include "theory/quantifiers/sygus/synth_conjecture.h"
+
+using namespace CVC4;
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+ExampleEvalCache::ExampleEvalCache(TermDbSygus* tds,
+                                   SynthConjecture* p,
+                                   Node f,
+                                   Node e)
+    : d_tds(tds), d_stn(e.getType())
+{
+  ExampleInfer* ei = p->getExampleInfer();
+  Assert(ei->hasExamples(f));
+  for (unsigned i = 0, nex = ei->getNumExamples(f); i < nex; i++)
+  {
+    std::vector<Node> input;
+    ei->getExample(f, i, input);
+    d_examples.push_back(input);
+  }
+  d_indexSearchVals = !d_tds->isVariableAgnosticEnumerator(e);
+}
+
+ExampleEvalCache::~ExampleEvalCache() {}
+
+Node ExampleEvalCache::addSearchVal(Node bv)
+{
+  if (!d_indexSearchVals)
+  {
+    // not indexing search values
+    return Node::null();
+  }
+  std::vector<Node> vals;
+  evaluateVec(bv, vals, true);
+  Trace("sygus-pbe-debug") << "Add to trie..." << std::endl;
+  Node ret = d_trie.addOrGetTerm(bv, vals);
+  Trace("sygus-pbe-debug") << "...got " << ret << std::endl;
+  // Clean up the cache data if necessary: if the enumerated term
+  // is redundant, its cached data will not be used later and thus should
+  // be cleared now.
+  if (ret != bv)
+  {
+    // immediately clear it
+    Trace("sygus-pbe-debug") << "...clear example cache" << std::endl;
+    clearEvaluationCache(bv);
+  }
```

Comment:
You could change this so that you just add to the cache if you have different results. It'd save doing this erasure operation. You will have the term and the evaluations anyway...

Reply:
Not sure it makes a difference, but I've changed it now.